rama-net/src/client/pool.rs

```diff
@@ -0,0 +1,577 @@
+            }),
+        }
+    }
+}
+
+/// Return type of [`Pool::get_connection_or_create_cb()`] to support advanced use cases
+pub enum GetConnectionOrCreate<C, F>
+where
+    F: FnOnce(C) -> LeasedConnection<C>,
+{
+    /// Connection was found in pool for given hash and is ready to be used
+    LeasedConnection(LeasedConnection<C>),
+    /// Pool doesn't have connection for hash but instead returns a function
+    /// which should be called by the external user to put a new connection
+    /// inside the pool. This fn also instantly returns a [`LeasedConnection`]
+    /// that is ready to be used
+    AddConnection(F),
+}
+
+impl<C> Pool<C> {
+    pub fn new(max_active: NonZeroU16, max_total: NonZeroU16) -> Result<Self, OpaqueError> {
+        if max_active > max_total {
+            return Err(OpaqueError::from_display(
+                "max_active should be <= then max_total connection",
+            ));
+        }
+        Ok(Self {
+            inner: Arc::new(PoolInner {
+                total_slots: Arc::new(Semaphore::new(Into::<u16>::into(max_total).into())),
+                active_slots: Arc::new(Semaphore::new(Into::<u16>::into(max_active).into())),
+                connections: Mutex::new(VecDeque::with_capacity(
+                    Into::<u16>::into(max_total).into(),
+                )),
+            }),
+        })
+    }
+
+    /// Get connection or create a new using provided async fn if we don't find one inside pool
+    pub async fn get_connection_or_create<F, Fut, E>(
+        &self,
+        hash: &str,
+        create_conn: F,
+    ) -> Result<LeasedConnection<C>, OpaqueError>
+    where
+        F: FnOnce() -> Fut,
+        Fut: Future<Output = Result<C, OpaqueError>>,
+    {
+        match self.get_connection_or_create_cb(hash).await? {
+            GetConnectionOrCreate::LeasedConnection(leased_connection) => Ok(leased_connection),
+            GetConnectionOrCreate::AddConnection(add) => {
+                let conn = create_conn().await?;
+                Ok(add(conn))
+            }
+        }
+    }
+
+    /// Get connection from pool or return fn to add a new one. See [`GetConnectionOrCreate`] for more info
+    pub async fn get_connection_or_create_cb(
+        &self,
+        hash: &str,
+    ) -> Result<GetConnectionOrCreate<C, impl FnOnce(C) -> LeasedConnection<C>>, OpaqueError> {
+        let active_permit = self
+            .inner
+            .active_slots
+            .clone()
+            .acquire_owned()
+            .await
+            .context("failed to acquire active slot permit")?;
+
+        let active_slot = ActiveSlot(active_permit);
+        let pool = Arc::<PoolInner<C>>::downgrade(&self.inner);
+
+        // Check if we can reuse stored connection
+        let pooled_conn = {
+            let mut connections = self.inner.connections()?;
+            connections
+                .iter()
+                .position(|stored| stored.hash == hash)
```

Comment:
In the end the only thing you really require it seems is that hash is `PartialEq` from what I can see. Perhaps also some other stuff for debugging e.g. fmt::Debug.

However might awnt to make that a generic so it doesn't have to be a string, as one might as well just use a number, some rich ID, a tuple or anything else. As long as it can be compared :)

Reply:
I would btw not call it hash but instead `ConnectionId` or `ConnectionKey`, as the user isn't forced to make a hash here.